File: src/ios/DBMeterPlugin.swift
import Foundation
import AVFoundation

/**
 * DBMeterPlugin - Cordova plugin to get decibel (dB) values from the microphone.
 *
 * This plugin provides the decibel level from the microphone for sound meter applications.
 */
@objc(DBMeterPlugin) class DBMeterPlugin: CDVPlugin {

    private let LOG_TAG = "DBMeterPlugin"
    private let SAMPLE_INTERVAL_MS = 100

    private var isListening: Bool = false
    private var audioRecorder: AVAudioRecorder!
    private var command: CDVInvokedUrlCommand!
    private var timer: DispatchSourceTimer!
    private var isTimerExists: Bool = false

    /**
     * Initialize with command delegate
     */
    init(commandDelegate: CDVCommandDelegate) {
        super.init()
        self.commandDelegate = commandDelegate
    }

    override init() {
        super.init()
    }

    /**
     * Permits to free the memory from the audioRecord instance
     */
    @objc(destroy:)
    func destroy(command: CDVInvokedUrlCommand) {
        if (self.isListening) {
            self.timer.suspend()
            self.isListening = false
        }

        self.command = nil

        if (self.audioRecorder != nil) {
            self.audioRecorder.stop()
            self.audioRecorder = nil

            let pluginResult = CDVPluginResult(status: .ok)
            self.commandDelegate!.send(pluginResult, callbackId: command.callbackId)
        } else {
            self.sendPluginError(callbackId: command.callbackId, errorCode: PluginError.DBMETER_NOT_INITIALIZED, errorMessage: "DBMeter is not initialized")
        }
    }

    /**
     * Initialize the audio recorder if not already initialized
     */
    private func initAudioRecorder() throws {
        if (self.audioRecorder == nil) {
            let url: URL = URL.init(fileURLWithPath: "/dev/null")

            let settings: [String: Any] = [
                AVFormatIDKey: Int(kAudioFormatAppleLossless),
                AVSampleRateKey: 44100.0,
                AVNumberOfChannelsKey: 1 as NSNumber,
                AVEncoderAudioQualityKey: AVAudioQuality.high.rawValue
            ]

            let audioSession: AVAudioSession = AVAudioSession.sharedInstance()
            try audioSession.setCategory(AVAudioSession.Category.playAndRecord)
            try audioSession.setActive(true)

            self.audioRecorder = try AVAudioRecorder(url: url, settings: settings)
            self.audioRecorder.isMeteringEnabled = true
        }
    }

    /**
     * Calculate the dB level from the audio recorder
     */
    private func calculateDbLevel() -> Int32 {
        self.audioRecorder.updateMeters()
        let peakPowerForChannel = pow(10, (self.audioRecorder.averagePower(forChannel: 0) / 20))
        let db = Int32(round(20 * log10(peakPowerForChannel) + 90))
        return db
    }

    /**
     * Starts listening the audio signal and sends dB values as a
     * CDVPluginResult keeping the same callback alive.
     */
    @objc(start:)
    func start(command: CDVInvokedUrlCommand) {
        self.commandDelegate!.run(inBackground: {
            self.command = command

            if (!self.isTimerExists) {
                self.timer = DispatchSource.makeTimerSource(queue: DispatchQueue.main)
                self.timer.schedule(deadline: .now(), repeating: DispatchTimeInterval.milliseconds(self.SAMPLE_INTERVAL_MS))
                self.timer.setEventHandler(handler: self.timerCallBack)
                self.isTimerExists = true
            }

            do {
                try self.initAudioRecorder()
            } catch {
                self.sendPluginError(callbackId: command.callbackId, errorCode: PluginError.DBMETER_NOT_INITIALIZED, errorMessage: "Error while initializing DBMeter")
                return
            }

            if (!self.isListening) {
                self.isListening = true
                self.audioRecorder.record()
                self.timer.resume()
            }
        })
    }

    /**
     * Get a single dB level reading (Promise-based)
     */
    @objc(getDbLevel:)
    func getDbLevel(command: CDVInvokedUrlCommand) {
        self.commandDelegate!.run(inBackground: {
            do {
                try self.initAudioRecorder()
            } catch {
                self.sendPluginError(callbackId: command.callbackId, errorCode: PluginError.DBMETER_NOT_INITIALIZED, errorMessage: "Error while initializing DBMeter")
                return
            }

            let wasListening = self.isListening
            if (!self.isListening) {
                self.audioRecorder.record()
            }

            // Give it a moment to get a reading
            Thread.sleep(forTimeInterval: 0.1)

            let db = self.calculateDbLevel()

            // Stop if it wasn't listening before
            if (!wasListening) {
                self.audioRecorder.stop()
            }

            let pluginResult = CDVPluginResult(status: .ok, messageAs: db)
            self.commandDelegate!.send(pluginResult, callbackId: command.callbackId)
        })
    }

    /**
     * Stops listening the audio signal.
     * Even if stopped, the AVAudioRecorder instance still exists.
     * To destroy this instance, please use the destroy method.
     */
    @objc(stop:)
    public func stop(command: CDVInvokedUrlCommand) {
        self.commandDelegate!.run(inBackground: {
            if (self.isListening) {
                self.isListening = false

                if (self.audioRecorder != nil && self.audioRecorder.isRecording) {
                    self.timer.suspend()
                    self.audioRecorder.stop()
                }

                let pluginResult = CDVPluginResult(status: .ok)
                self.commandDelegate!.send(pluginResult, callbackId: command.callbackId)
            } else {
                self.sendPluginError(callbackId: command.callbackId, errorCode: PluginError.DBMETER_NOT_LISTENING, errorMessage: "DBMeter is not listening")
            }
        })
    }

    /**
     * Returns whether the DBMeter is listening.
     */
    @objc(isListening:)
    func isListening(command: CDVInvokedUrlCommand?) -> Bool {
        if (command != nil) {
            let pluginResult = CDVPluginResult(status: .ok, messageAs: self.isListening)
            self.commandDelegate!.send(pluginResult, callbackId: command!.callbackId)
        }
        return self.isListening
    }

    /**
     * Timer callback for continuous monitoring
     */
    private func timerCallBack() {
        autoreleasepool {
            if (self.isListening && self.audioRecorder != nil) {
                let db = self.calculateDbLevel()

                let pluginResult = CDVPluginResult(status: .ok, messageAs: db)
                pluginResult.setKeepCallbackAs(true)
                self.commandDelegate!.send(pluginResult, callbackId: self.command.callbackId)
            }
        }
    }

    /**
     * Convenient method to send plugin errors.
     */
    private func sendPluginError(callbackId: String, errorCode: PluginError, errorMessage: String) {
        let pluginResult = CDVPluginResult(status: .error, messageAs: ["code": errorCode.hashValue, "message": errorMessage])
        self.commandDelegate!.send(pluginResult, callbackId: callbackId)
    }

    enum PluginError: String {
        case DBMETER_NOT_INITIALIZED = "0"
        case DBMETER_NOT_LISTENING = "1"
    }
}

File: src/android/DBMeterPlugin.java
package dbmeterplugin;

import android.Manifest;
import android.media.AudioFormat;
import android.media.AudioManager;
import android.media.AudioRecord;
import android.media.AudioTrack;
import android.media.MediaRecorder;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.LOG;
import org.apache.cordova.PluginResult;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.Timer;
import java.util.TimerTask;

/**
 * DBMeterPlugin - Cordova plugin to get decibel (dB) values from the microphone.
 *
 * This plugin provides the decibel level from the microphone for sound meter applications.
 */
public class DBMeterPlugin extends CordovaPlugin {

    private static final String LOG_TAG = "DBMeterPlugin";
    private static final int REQ_CODE = 0;
    private static final int SAMPLE_INTERVAL_MS = 100;

    private AudioRecord audioRecord;
    private short[] buffer;
    private Timer timer;
    private boolean isListening = false;
    private CallbackContext startCallbackContext;

    @Override
    public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {
        if (!hasRecordAudioPermission()) {
            requestRecordAudioPermission();
            return true;
        }

        switch (action) {
            case "start":
                this.start(callbackContext);
                return true;
            case "getDbLevel":
                this.getDbLevel(callbackContext);
                return true;
            case "stop":
                this.stop(callbackContext);
                return true;
            case "isListening":
                this.isListening(callbackContext);
                return true;
            case "destroy":
                this.destroy(callbackContext);
                return true;
            default:
                LOG.e(LOG_TAG, "Not a valid action: " + action);
                return false;
        }
    }

    /**
     * Check if RECORD_AUDIO permission is granted
     */
    private boolean hasRecordAudioPermission() {
        return cordova.hasPermission(Manifest.permission.RECORD_AUDIO);
    }

    /**
     * Request RECORD_AUDIO permission
     */
    private void requestRecordAudioPermission() {
        cordova.requestPermission(this, REQ_CODE, Manifest.permission.RECORD_AUDIO);
    }

    /**
     * Permits to free the memory from the audioRecord instance
     *
     * @param callbackContext The callback context used when calling back into JavaScript.
     */
    public void destroy(CallbackContext callbackContext) {
        if (this.audioRecord != null) {
            this.isListening = false;
            this.audioRecord.stop();
            this.audioRecord.release();
            this.audioRecord = null;
            if (this.timer != null) {
                this.timer.cancel();
                this.timer = null;
            }
            callbackContext.success();
        } else {
            sendPluginError(callbackContext, PluginError.DBMETER_NOT_INITIALIZED, "DBMeter is not initialized");
        }
    }

    /**
     * Initialize the audio recorder if not already initialized
     */
    private void initAudioRecorder() {
        if (this.audioRecord == null) {
            int rate = AudioTrack.getNativeOutputSampleRate(AudioManager.STREAM_SYSTEM);
            int bufferSize = AudioRecord.getMinBufferSize(rate, AudioFormat.CHANNEL_IN_MONO, AudioFormat.ENCODING_PCM_16BIT);

            this.audioRecord = new AudioRecord(
                    MediaRecorder.AudioSource.MIC,
                    rate,
                    AudioFormat.CHANNEL_IN_MONO,
                    AudioFormat.ENCODING_PCM_16BIT,
                    bufferSize);

            this.buffer = new short[bufferSize];
        }
    }

    /**
     * Calculate the dB level from the audio buffer
     *
     * @return The calculated dB value
     */
    private double calculateDbLevel() {
        int readSize = this.audioRecord.read(this.buffer, 0, this.buffer.length);
        double maxAmplitude = 0;

        for (int i = 0; i < readSize; i++) {
            if (Math.abs(this.buffer[i]) > maxAmplitude) {
                maxAmplitude = Math.abs(this.buffer[i]);
            }
        }

        double db = 0;
        if (maxAmplitude != 0) {
            db = 20.0 * Math.log10(maxAmplitude / 32767.0) + 90;
        }

        return db;
    }

    /**
     * Starts listening the audio signal and sends dB values as a
     * {@link org.apache.cordova.PluginResult PluginResult} using a
     * {@link java.util.TimerTask TimerTask}.
     *
     * @param callbackContext The callback context used when calling back into JavaScript.
     */
    public void start(final CallbackContext callbackContext) {
        this.startCallbackContext = callbackContext;
        final DBMeterPlugin that = this;

        cordova.getThreadPool().execute(new Runnable() {
            public void run() {
                that.initAudioRecorder();

                if (!that.isListening) {
                    that.isListening = true;
                    that.audioRecord.startRecording();

                    that.timer = new Timer(LOG_TAG, true);

                    TimerTask timerTask = new TimerTask() {
                        public void run() {
                            if (that.isListening && that.audioRecord != null) {
                                double db = that.calculateDbLevel();
                                LOG.d(LOG_TAG, Double.toString(db));

                                PluginResult result = new PluginResult(PluginResult.Status.OK, (float) db);
                                result.setKeepCallback(true);
                                callbackContext.sendPluginResult(result);
                            }
                        }
                    };
                    that.timer.scheduleAtFixedRate(timerTask, 0, SAMPLE_INTERVAL_MS);
                }
            }
        });
    }

    /**
     * Get a single dB level reading (Promise-based)
     *
     * @param callbackContext The callback context used when calling back into JavaScript.
     */
    public void getDbLevel(final CallbackContext callbackContext) {
        final DBMeterPlugin that = this;

        cordova.getThreadPool().execute(new Runnable() {
            public void run() {
                that.initAudioRecorder();

                boolean wasListening = that.isListening;
                if (!that.isListening) {
                    that.audioRecord.startRecording();
                }

                // Give it a moment to get a reading
                try {
                    Thread.sleep(100);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }

                double db = that.calculateDbLevel();

                // Stop if it wasn't listening before
                if (!wasListening) {
                    that.audioRecord.stop();
                }

                PluginResult result = new PluginResult(PluginResult.Status.OK, (float) db);
                callbackContext.sendPluginResult(result);
            }
        });
    }

    /**
     * Stops listening the audio signal.
     * Even if stopped, the {@link android.media.AudioRecord AudioRecord} instance still exists.
     * To destroy this instance, please use the {@link #destroy(CallbackContext) destroy} method.
     *
     * @param callbackContext The callback context used when calling back into JavaScript.
     */
    public void stop(final CallbackContext callbackContext) {
        final DBMeterPlugin that = this;

        cordova.getThreadPool().execute(new Runnable() {
            public void run() {
                if (that.isListening) {
                    that.isListening = false;

                    if (that.timer != null) {
                        that.timer.cancel();
                        that.timer = null;
                    }
                    if (that.audioRecord != null) {
                        that.audioRecord.stop();
                    }
                    callbackContext.success();
                } else {
                    sendPluginError(callbackContext, PluginError.DBMETER_NOT_LISTENING, "DBMeter is not listening");
                }
            }
        });
    }

    /**
     * Returns whether the DBMeter is listening.
     *
     * @param callbackContext The callback context used when calling back into JavaScript.
     */
    public void isListening(final CallbackContext callbackContext) {
        PluginResult result = new PluginResult(PluginResult.Status.OK, this.isListening);
        callbackContext.sendPluginResult(result);
    }

    /**
     * Convenient method to send plugin errors.
     *
     * @param callbackContext The callback context used when calling back into JavaScript.
     * @param error           The error code to return
     * @param message         The error message to return
     */
    private void sendPluginError(CallbackContext callbackContext, PluginError error, String message) {
        JSONObject jsonObject = new JSONObject();
        try {
            jsonObject.put("code", error.ordinal());
            jsonObject.put("message", message);
        } catch (JSONException e) {
            e.printStackTrace();
        }
        callbackContext.error(jsonObject);
    }

    public enum PluginError {
        DBMETER_NOT_INITIALIZED,
        DBMETER_NOT_LISTENING
    }
}
